File: app/src/main/java/s15666/pjwstk/pamobmi/ui/calories/CaloriesViewModel.java
package s15666.pjwstk.pamobmi.ui.calories;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import s15666.pjwstk.pamobmi.benedictharris.EnergyExpenditure;
import s15666.pjwstk.pamobmi.benedictharris.Gender;

public class CaloriesViewModel extends ViewModel {

    private final MutableLiveData<Gender> gender = new MutableLiveData<>();
    private final MutableLiveData<Integer> age = new MutableLiveData<>();
    private final MutableLiveData<EnergyExpenditure> energyExpenditure = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isValid = new MutableLiveData<>();

    public CaloriesViewModel() {
        gender.setValue(Gender.MALE);
        isValid.setValue(Boolean.FALSE);
        energyExpenditure.setValue(EnergyExpenditure.LIGHT);
    }

    MutableLiveData<Gender> getGender() {
        return gender;
    }

    void setGender(Gender gender) {
        this.gender.setValue(gender);
        checkValid();
    }

    MutableLiveData<Integer> getAge() {
        return age;
    }

    void setAge(Integer age) {
        this.age.setValue(age);
        checkValid();
    }

    MutableLiveData<EnergyExpenditure> getEnergyExpenditure() {
        return energyExpenditure;
    }

    void setEnergyExpenditure(EnergyExpenditure e) {
        energyExpenditure.setValue(e);
        Boolean isValid = e != null;
        if(isValid != this.isValid.getValue()) {
            this.isValid.setValue(isValid);
        }
    }

    private void checkValid() {
        Boolean newValue = age.getValue() != null && age.getValue() > 0;
        if(isValid.getValue() != newValue) {
            isValid.setValue(newValue);
        }
    }

    MutableLiveData<Boolean> isValid() {
        return isValid;
    }

}

File: app/src/main/java/s15666/pjwstk/pamobmi/ui/calories/CaloriesViewModel.kt
package s15666.pjwstk.pamobmi.ui.calories

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import s15666.pjwstk.pamobmi.benedictharris.EnergyExpenditure
import s15666.pjwstk.pamobmi.benedictharris.Gender

class CaloriesViewModel : ViewModel() {
    val gender = MutableLiveData<Gender>()
    val age = MutableLiveData<Int?>()
    val energyExpenditure = MutableLiveData<EnergyExpenditure>()
    private val isValid = MutableLiveData<Boolean>()

    init {
        gender.value = Gender.MALE
        isValid.value = java.lang.Boolean.FALSE
        energyExpenditure.value = EnergyExpenditure.LIGHT
    }

    fun setGender(gender: Gender) {
        this.gender.value = gender
    }

    fun setAge(age: Int) {
        this.age.value = age
        if (age > 0) updateValid(true) else updateValid(false)
    }

    private fun updateValid(valid: Boolean) { if (valid != isValid.value) isValid.value = valid }

    fun isValid(): Boolean = isValid.value!!

}